### PR TITLE
Switch to using MXNet 1.3 for migration and add Keras integ test

### PR DESCRIPTION
**Pull request by laurenyu**
*Description of changes:*
There are a few things going on here:
1. Add Dockerfiles for 1.3.0 that use the new script mode/SageMaker Containers stuff
1. Revert the changes to the 1.2.1 Dockerfiles as those will not change with the upcoming release
1. Add an integration test for training with Keras, a feature that was released with MXNet 1.3

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docker/1.3.0/final/Dockerfile.cpu [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# Validate that arguments are specified
+RUN test $framework_installable || exit 1 && \
+    test $py_version || exit 1
+
+WORKDIR /tmp
+
+RUN apt-get update && \
+    apt-get -y install build-essential libopencv-dev libopenblas-dev libjemalloc-dev libgfortran3 \
+    python-dev python3-dev wget curl nginx
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.cpu]: Given that you are not building anything, change {libname}-dev to just {libname}. It saves a lot of image size. Remove build-essential as well.

File: docker/1.3.0/final/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+# Validate that arguments are specified
+RUN test $framework_installable || exit 1 && \
+    test $py_version || exit 1
+
+WORKDIR /tmp
+
+RUN apt-get update && \
+    apt-get -y install build-essential libopencv-dev libatlas-base-dev libcurl4-openssl-dev libgtest-dev \
+        libjemalloc-dev python-dev python3-dev unzip git wget curl nginx
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.gpu]: Same here

File: docker/1.3.0/final/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    framework_support_installable_local=$(basename $framework_support_installable) && \
+    \
+    pip install $framework_installable_local && \
+    pip install $framework_support_installable_local && \
+    \
+    rm $framework_installable_local && \
+    rm $framework_support_installable_local
+
+RUN pip install onnx \
+                keras-mxnet
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.gpu]: Define specific version. Consider adding these ti setup.py

**Comment by laurenyu** [comment]
removed "-dev" from python in the apt-get lines, added versions to onnx and keras-mxnet in the pip install lines. ONNX test will come later

File: docker/1.3.0/final/Dockerfile.cpu [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# This is here to make our installed version of OpenCV work.
+# https://stackoverflow.com/questions/29274638/opencv-libdc1394-error-failed-to-initialize-libdc1394
+# TODO: Should we be installing OpenCV in our image like this? Is there another way we can fix this?
+RUN ln -s /dev/null /dev/raw1394
+
+ENV PYTHONDONTWRITEBYTECODE=1 \
+    PYTHONUNBUFFERED=1 \
+    LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:/usr/local/lib"
+
+ENV SAGEMAKER_TRAINING_MODULE sagemaker_mxnet_container.training:main
```
Comment (ChoiByungWook) [inline docker/1.3.0/final/Dockerfile.cpu]: is the serving env being added later?

Comment (laurenyu) [reply docker/1.3.0/final/Dockerfile.cpu]: yes, once the implementation that supports it is complete :)

File: docker/1.2.1/final/Dockerfile.gpu [hunk 5]
```diff
@@ -1,7 +1,7 @@
 FROM nvidia/cuda:9.0-cudnn7-devel
 ARG framework_installable
 ARG py_version
-ARG framework_support_installable=sagemaker_mxnet_container-2.0.0.tar.gz
```
[the remark this answers is not in the document]
Comment (mvsusp) [reply docker/1.2.1/final/Dockerfile.gpu]: On line 12 you can remove build-essential and please check the necessity of libopencv-dev libatlas-base-dev libcurl4-openssl-dev libgtest-dev and libjemalloc-dev

I am not sure you need python or python-dev

Change python3-dev to python3

File: docker/1.2.1/final/Dockerfile.gpu [hunk 6]
```diff
@@ -1,7 +1,7 @@
...
+ARG framework_support_installable=sagemaker_mxnet_container-1.0.0.tar.gz
```
Comment (mvsusp) [inline docker/1.2.1/final/Dockerfile.gpu]: On line 21, please delete get-pip.py after installing it.

File: docker/1.3.0/final/Dockerfile.gpu [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+# Validate that arguments are specified
+RUN test $framework_installable || exit 1 && \
+    test $py_version || exit 1
+
+WORKDIR /tmp
+
+RUN apt-get update && \
+    apt-get -y install libopencv-dev libatlas-base-dev libcurl4-openssl-dev libgtest-dev \
+        libjemalloc-dev python python3 unzip git wget curl nginx
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.gpu]: same here

File: docker/1.2.1/final/Dockerfile.gpu (hunk 5 above)
[the remark this answers is not in the document]
Comment (laurenyu) [reply docker/1.2.1/final/Dockerfile.gpu]: I wasn't able to find anything for libopencv, etc. without "-dev" - the image build exited with an error

removed -dev(el) on the others

**Comment by laurenyu** [comment]
@mvsusp removed some of the libraries and changed cuda:9.0-cudnn7-devel to cuda:9.0-cudnn7-runtime. I'm not updating the 1.2.1 Dockerfiles because this branch is for 1.3.0, but if it seems important enough, we can update 1.2.1 in the master branch of this repo

File: docker/1.3.0/final/Dockerfile.cpu [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ARG framework_support_installable=sagemaker_mxnet_container-2.0.0.tar.gz
+
+# Validate that arguments are specified
+RUN test $framework_installable || exit 1 && \
+    test $py_version || exit 1
+
+WORKDIR /tmp
+
+RUN apt-get update && \
+    apt-get -y install libopencv-dev libopenblas-dev python python3 wget curl nginx
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.cpu]: always end a apt-get layer with \
 && rm -rf /var/lib/apt/lists/*

https://docs.docker.com/v17.09/engine/userguide/eng-image/dockerfile_best-practices/#run

Comment (mvsusp) [reply docker/1.3.0/final/Dockerfile.cpu]: Install only one python version. You can use a variable for that https://github.com/aws/sagemaker-tensorflow-container/blob/script-mode/docker/1.10.0/Dockerfile.cpu#L45

File: docker/1.3.0/final/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+# Symlink /usr/bin/python to the python version we're building for.
+RUN rm /usr/bin/python && ln -s "/usr/bin/python$py_version" /usr/bin/python
+
+# Install pip
+RUN curl -O https://bootstrap.pypa.io/get-pip.py && \
+    /usr/bin/python get-pip.py && \
+    rm get-pip.py
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.cpu]: Install pip like this https://github.com/aws/sagemaker-tensorflow-container/blob/master/docker/1.11.0/final/py2/Dockerfile.cpu#L19

File: docker/1.3.0/final/Dockerfile.cpu [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+RUN framework_installable_local=$(basename $framework_installable) && \
+    framework_support_installable_local=$(basename $framework_support_installable) && \
+    \
+    pip install $framework_installable_local && \
+    pip install $framework_support_installable_local && \
+    \
+    rm $framework_installable_local && \
+    rm $framework_support_installable_local
+
+RUN pip install onnx==1.2.1 \
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.cpu]: Less layers is better. Put all pip installs in the same layer.

File: docker/1.3.0/final/Dockerfile.gpu [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+# Validate that arguments are specified
+RUN test $framework_installable || exit 1 && \
+    test $py_version || exit 1
+
+WORKDIR /tmp
+
+RUN apt-get update && \
+    apt-get -y install libopencv-dev libatlas-base-dev libcurl4-openssl-dev \
+        python python3 unzip git wget curl nginx
```
Comment (mvsusp) [inline docker/1.3.0/final/Dockerfile.gpu]: remove unzip wget curl

File: docker/1.3.0/final/Dockerfile.cpu (hunk 8 above)
Comment (mvsusp) [reply docker/1.3.0/final/Dockerfile.cpu]: Add --no-install-recommends

**Pull request merged**